### PR TITLE
Training code

### PR DESCRIPTION
**Pull request by AvivNavon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,7 +35,88 @@ pip install -e .
...
+]
+```
+
+An example of a dataset can be found in [Voxpopuli-NER-EN](https://huggingface.co/datasets/aiola/Voxpopuli_NER).
+
+## ## Training and finetuning
+
+To train the model, run the following command (or modify it according to your needs):
+
+python trainer.py --audio-root-dir /home/ec2-user/speech_data/whisperner --batch-size=2 --entity-dropout-prob=0.1 --eval-steps=10000 --exp-name=nuner-large-250k-350k-2-negative-samples --fp16=False --gradient-accumulation-steps=2 --lora-merge-and-unload=False --lr=1e-06 --max-eval-samples=1000 --max-steps=250000 --n-neg-samples=2 --output-path=../artifacts/tmp --parts-to-freeze=encoder --predict-with-generate=False --save-steps=10000 --test-data-path /home/ec2-user/speech_data/whisperner/CrossNER_AI/ner_test_fix.json  --train-data-path /home/ec2-user/speech_data/whisperner/CrossNER_AI/ner_test_fix.json --use-lora=False --validation-data-path /home/ec2-user/speech_data/whisperner/CrossNER_AI/ner_test_fix.json --wandb-logging=false --whisper-model-name=openai/whisper-large-v2
```
Comment (AvivSham) [inline README.md:90]:
```suggestion
```

redundant isn't it?

File: whisper_ner/dataset.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+from typing import Dict, List, Union
+
+import torch
+import torchaudio
+from transformers import WhisperProcessor
+
+
+from whisper_ner.utils import set_logger, token_padding
+
+set_logger()
```
Comment (AvivSham) [inline whisper_ner/dataset.py:15]:
```suggestion
from whisper_ner.utils import token_padding
```

The logger will be set from `trainer.py` no?

File: whisper_ner/dataset.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            dataset = json.load(f)
+
+        if max_samples is not None:
+            logging.info(f"Using only {max_samples} samples from the dataset.")
+            dataset = dataset[:max_samples]
+
+        self.dataset = dataset
+        all_ner_tags = [
+            tag for d in dataset
+            for _, _, tag, _, _ in [self._unify_ner_tag(t) for t in d["ner"]]
```
Comment (AvivSham) [inline whisper_ner/dataset.py:44]: looks a bit ugly instead of `_, _, tag, _, _` use single param and slice using [2] index.

File: whisper_ner/dataset.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        array = resampler(speech_array)[0].numpy()
+        return self.processor.feature_extractor(
+            array, sampling_rate=resampling_to
+        ).input_features[0]
+
+    def _sample_negative_ner_tags(self, idx):
+        query_record = self.dataset[idx]
+        query_ner_tags = query_record["ner"]
+        query_ner_tags = [self._unify_ner_tag(tag) for tag in query_ner_tags]
+        query_ner_types = list(set([tag for _, _, tag, _, _ in query_ner_tags]))
```
Comment (AvivSham) [inline whisper_ner/dataset.py:100]: +1

File: README.md [hunk 5]
```diff
@@ -34,7 +35,88 @@ pip install -e .
...
+  --output-path <output-dir> \
+  --parts-to-freeze encoder \
+  --predict-with-generate False \
+  --audio-root-dir <root-of-audio-file> \
+  --test-data-path <test-json-path> \
+  --train-data-path <train-json-path> \
+  --validation-data-path <val-json-path> \
+  --wandb-logging true \
+  --exp-name <wandb-exp-name> \
+  --wandb-entity <wandb-entity>
```
Comment (AvivSham) [inline README.md:116]: consider adding info about the args

File: whisper_ner/trainer.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    if args_i.train:
+        if not args_i.use_lora:
+            model.freeze_model_parts(args_i.parts_to_freeze)
+        trainer.train()
+
+    if args_i.use_lora and args_i.lora_merge_and_unload:
+        model = model.merge_and_unload()
+
+    # todo: save tokenizer + processor to separate dir
```
Comment (AvivSham) [inline whisper_ner/trainer.py:84]:
```suggestion
```

File: whisper_ner/trainer.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    model.save_pretrained(model_comp_path_str)
+    processor.tokenizer.save_pretrained(model_comp_path_str)
+    processor.save_pretrained(model_comp_path_str)
+
+    results = trainer.evaluate(eval_dataset=dataset["test"])
+
+    message = f"loss: {results['eval_loss']}"
+    if args_i.compute_wer:
+        message += f", WER: {results['eval_wer']}"
+    logging.info(message)
```
Comment (AvivSham) [inline whisper_ner/trainer.py:98]: Lets add test prefix.

```suggestion
    message = f"test loss: {results['eval_loss']}"
    if args_i.compute_wer:
        message += f", test WER: {results['eval_wer']}"
    logging.info(message)
```

File: whisper_ner/utils/metrics.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    pred_ids = pred.predictions
+    label_ids = pred.label_ids
+
+    label_ids[label_ids == -100] = tokenizer.pad_token_id
+
+    pred_str = tokenizer.batch_decode(pred_ids, skip_special_tokens=True)
+    label_str = tokenizer.batch_decode(label_ids, skip_special_tokens=True)
+
+    wer, _ = compute_wer(pred_str, label_str)
+    return dict(wer=wer)
```
Comment (AvivSham) [inline whisper_ner/utils/metrics.py:79]:
```suggestion
```

File: whisper_ner/trainer.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        args_i.whisper_model_name, language=args_i.language, task="transcribe"
+    )
+    dataset = get_dataset(args_i, processor)
+    data_collator = DataCollatorSeq2SeqWithPadding(processor)
+
+    compute_metrics_arg = (
+        partial(compute_metrics, tokenizer=processor.tokenizer)
+        if args_i.compute_wer
+        else None
+    )
```
Comment (AvivSham) [inline whisper_ner/trainer.py:64]:
```suggestion
```

File: whisper_ner/trainer.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    trainer = Seq2SeqTrainer(
+        model=model,
+        args=training_args,
+        data_collator=data_collator,
+        train_dataset=dataset["train"],
+        eval_dataset={"validation": dataset["validation"]},
+        tokenizer=processor.feature_extractor,
+        compute_metrics=compute_metrics_arg,
```
Comment (AvivSham) [inline whisper_ner/trainer.py:73]:
```suggestion
```

**Review by AvivSham — CHANGES_REQUESTED** [review]
See inline

File: whisper_ner/trainer.py (hunk 9 above)
Comment (AvivNavon) [reply whisper_ner/trainer.py:64]: Why removing this?

Comment (AvivSham) [reply whisper_ner/trainer.py:64]: We do not use this we only use the loss for monitoring.

File: whisper_ner/dataset.py [hunk 11]
```diff
@@ -36,12 +35,12 @@ def __init__(
 
         if max_samples is not None:
             logging.info(f"Using only {max_samples} samples from the dataset.")
-            dataset = dataset[:max_samples]
+            dataset = dataset.sample(frac=1.)[:max_samples]
```
Comment (AvivSham) [inline whisper_ner/dataset.py:38]: n=max_samples for efficient.

**Review by AvivSham — APPROVED** [review]
